### ISSUE TITLE
Remove need for `addError`, `removeError`, ...

### ISSUE THREAD

**Issue by marcandre** [guillaumepotier/Parsley.js#1073]
With the right set of feature, there shouldn't be a need for `addError`, `updateError` and `removeError`.

Seems to be causes for this need:

[] custom error messages from validators

Others?

**Comment by marcandre** [comment]
Other reasons invoked:

> 2) Validation that requires more than one field ("whole form" validation), or Integration with existing code that does not work per-field.

I've added [an example](http://parsleyjs.org/doc/examples/custom-validator-events.html) of such validations.

> 3) has a different event model (async validations that are not covered by the model used in Parsley)

Not sure what these would be. Example anyone?

> 4) Multi-step validation, i.e. where one step must be completed (generating a token, hitting an API) before other validations can run

Couldn't these be combined using promises? I can imagine a case where two APIs must be called one after the other, but one could have a custom validator that returns

    return $.ajax('first-api', ...).then(function(result) {
        return $.ajax('second-api', ...);
    }

Or am I missing something?

**Comment by papiro** [comment]
What does the `return` statement inside of the `.then` do, besides making sure no code after it gets executed?  Is there a way to leverage the return value?

**Comment by marcandre** [comment]
@papiro Of course, that's how you "pipe" promises. The resulting promise will resolve only after the first and the second ajax calls resolved. The result will be the result of the last ajax call.

**Comment by dcarral** [comment]
Commenting in issue #1073 due after getting the following message in the console:

> Accessing ParsleyUI is deprecated. Call 'addError' on the instance directly. Please comment in issue 1073 as to your need to call this method.

The point is that if I change from:
```
window.ParsleyUI.addError(this, 'error-name', 'Message explaining the error');
```
to:
```
this.addError('error-name', 'Message explaining the error');
````
I get an `Uncaught TypeError: Cannot read property 'name' of undefined` message.

Fwiw, this code resides inside a Parsley event handler. @marcandre @guillaumepotier could you tell me what I'm doing wrong?

**Comment by marcandre** [comment]
You need to call `this.addError('error-name', {message: 'Message explaining the error'})`.

Could you please provide the reason why you are using `addError` at all? I'd very much like to get rid of any need people have of calling that.

**Comment by dcarral** [comment]
Thanks for the quick feedback @marcandre, appreciate it :) 

Regarding your question, the reason to add the errors manually is the lack of support for grouped select boxes validation. 

The use case consists on two select boxes (month & year) which should be validated:

 1. Independently (with default requireness validation)
 2. As a group (with custom validation)

In order to achieve the group validation, I'm currently hacking a solution consisting of adding & removing errors manually within Parsley event handlers. For what it's worth, in the middle of this (kind of funny) adventure, I've just run into issue #1085.

**Comment by marcandre** [comment]
@dcarral Please consider using a custom validator instead, otherwise it's not clear how things can remain consistent. See point 2 above.

**Comment by jesse-s** [comment]
> > 3) has a different event model (async validations that are not covered by the model used in Parsley)

> Not sure what these would be. Example anyone?

I have my own async logic, and I need serverside inline validation with different kinds of potential error messages and a success handler.

How do I apply `this.addError('error-name', {message: 'Message explaining the error'})` to specific field? Something like:
`$('#my-input').parsley().addError()`

**Comment by vedraan** [comment]
I'm using v2.3.11 and there appears to be no `addError()` method on the instance:
`$('#login-form').parsley().addError` is `undefined` so I was wondering how do I manually activate errors now since calling `window.ParsleyUI.addError` directs me to this issue. 

The docs seem to be out of date and still suggest using `window.ParsleyUI.addError`.

My use case is that I use an API which does not have a possibility to AJAX check if the a field is valid. I only get errors after POSTing the entire form, e.g. if I submit a register form the API might respond with:
```
{
  "message": "VALIDATION_ERROR",
  "body": [
    "error_fields",
    {
      "email": {
        "failed_rule": "Unique",
        "message": "The email has already been taken."
      },
      "password": {
        "failed_rule": "Complexity",
        "message": "The password is not complex enough."
      }
    }
  ]
}
```
Which is a great response which I can easily traverse and map the error message to the proper fields, but I can not see how that is possible with the latest Parsley version.

This is a very simple example, but I'm dealing with very complex CMS forms which can return errors for multiple fields with no way to validate all the error types client-side before submit.

I could of course display bulk server messages or render them manually in each error container but I would love using Parsley for this since I'm already using it for client side validation.

**Comment by marcandre** [comment]
@vedraan: `addError` need to be called on a parsley field instance, not a form instance... Thanks for your use case, it's a decent one.

**Comment by vedraan** [comment]
I was convinced I tried it on field instance as well, but probably missed something. Anyway if someone else finds this thread like me here's a little help on how to do it (at least until the docs get updated).
```JavaScript
$('[name="email"]').parsley().addError('error-1', { message: 'Message for error-1' }); // add error message
$('[name="email"]').parsley().addError('error-2', { message: 'Message for error-2' }); // add one more error message
$('[name="email"]').parsley().removeError('error-1'); // remove first message
```

**Comment by marcandre** [comment]
Doc updated, let me know if I missed something.

**Comment by ewerkema** [comment]
I agree with your point that the _custom validators_ should replace the functions 'addError', 'updateError' and 'removeError', but with the current _custom validators_ I am not able to implement it for all situations. For example:

1. When trying to check whether a username/password combination is correct. You do not want to constantly show the user whether they have typed the wrong/correct password. You would need to have a custom validator that is only being triggered on 'form:validated'.

```javascript
window.Parsley.addValidator('password', {
	validateString : function (value) {
		var password = $("#password").val();
		var result = getUser();
		return (result.length!==0 && hash.verify(password,result[0].user_password) );
	},
	messages : {
		en : lang("error_passwordincorrect")
	}
});
```

2. When retrieving data from a database using NodeJS. Because of the possible time delay from the database, the result is not yet returned and Parsley will show an error. Normally you would create a callback function that is being called after the result is returned. 

```javascript
window.Parsley.addValidator('emailName', {
  validateString: function(value, requirement) {
    var result;
    db.getOnlineQuery("getUserIdbyEmail", [value], function(rows) {
      // Callback from database
      result = rows;
    });
    return (result.length===0);
  },
  messages: {
    en: lang("error_emailnotunique")
  }
});
```

And maybe there are some more cases in which the _custom validators_ cannot be used. If I need to explain things better, please let me know.

**Comment by alniks** [comment]
I'm using Version 2.2.0 and there is no updateError or getMessages on parsley field instances.

**Comment by benjaminpick** [comment]
Note that a big difference between "custom validator" and "custom error message" is also that the error message is only the UI part. So if a field has a custom error message, the form validates and submits anyway - probably not what you'd want. (You could create custom validator that returns false if there is a custom error message, but well, why not implement the message there as well ...)

**Comment by benjaminpick** [comment]
> You can return $.Deferred.reject('My custom error') from a custom validator. Maybe I should accept returning a string directly too...

Can you please give me an example? `return $.Deferred.reject('My custom error');` tells me that reject is not a function.

**Comment by marcandre** [comment]
Sorry, edited. `$.Deferred().reject(...)`

**Comment by kojotkk** [comment]
If you remove possibility of addError and removeError usage, then there will be no clean way to add standard formatted message (when message should be added to data-parsley-error-container).

Example from parsley doc is here - http://parsleyjs.org/doc/examples/events.html. Adding message and toggling class is a bit ugly in this example.

There is possibility now to validate it through events and use addError to add standard looking error message.

Maybe doc is a bit outdated and currently there is a way to validate two fields (like with data-parsley-group) using custom validator?

**Comment by janwidmer** [comment]
I am having another reason why I cannot use Custom Validators:
My backend is not able to return Server Side Validation on a per field basis, but only for all fields. Therefore I am doing an ajax request to do the server side validation and then appending the errors to the different fields..
